src/Algebra/Solver/Monoid/Reflection.agda

```diff
@@ -0,0 +1,185 @@
+{-# OPTIONS --without-K --safe #-}
+
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Reflection-based solver for monoid equalities
+------------------------------------------------------------------------
+
+module Algebra.Solver.Monoid.Reflection where
+
+open import Algebra
+open import Function
+
+open import Data.Bool    as Bool    using (Bool; _∨_; if_then_else_)
+open import Data.Maybe   as Maybe   using (Maybe; just; nothing; maybe)
+open import Data.List    as List    using (List; _∷_; [])
+open import Data.Nat     as ℕ       using (ℕ; suc; zero)
+open import Data.Product as Product using (_×_; _,_)
+
+open import Agda.Builtin.Reflection
+
+----------------------------------------------------------------------
+-- The Expr type with homomorphism proofs
+----------------------------------------------------------------------
+
+infixl 7 _⊙_
+data Expr {a} (A : Set a) : Set a where
+  _⊙_ : Expr A → Expr A → Expr A
+  ⓔ  : Expr A
+  ⟦_↑⟧ : A → Expr A
+
+module _ {m₁ m₂} (mon : Monoid m₁ m₂) where
+  open Monoid mon
+
+  ⟦_↓⟧ : Expr Carrier → Carrier
+  ⟦ x ⊙ y  ↓⟧ = ⟦ x ↓⟧ ∙ ⟦ y ↓⟧
+  ⟦ ⓔ      ↓⟧ = ε
+  ⟦ ⟦ x ↑⟧ ↓⟧ = x
+
+  ⟦_⇓⟧′ : Expr Carrier → Carrier → Carrier
+  ⟦ x ⊙ y  ⇓⟧′ z = ⟦ x ⇓⟧′ (⟦ y ⇓⟧′ z)
+  ⟦ ⓔ      ⇓⟧′ y = y
+  ⟦ ⟦ x ↑⟧ ⇓⟧′ y = x ∙ y
+
+  ⟦_⇓⟧ : Expr Carrier → Carrier
+  ⟦ x ⇓⟧ = ⟦ x ⇓⟧′ ε
+
+  open import Relation.Binary.Reasoning.Setoid setoid
```

Comment:
Please could you import this at the top of the file (e.g. as `SetoidReasoning`) and then open it here.